src/model/File_data.C

```diff
@@ -32,31 +45,192 @@ Bytes::operator std::string() const
-{
-    return repr_trait::read(v, statement, column, size);
+int File_data::populate_contents() {
+
+    // Stringify assignment number, user, and data. 
+	const std::string user = file_meta_->submission()->user1()->name();
+    const std::string dir_path = CONFIG.gsc_repo + "assignment_" + 
+			std::to_string(file_meta_->submission()->assignment_number()) +  
+			"/" + user;
+
+	const std::string file_path = dir_path + '/' + file_meta_->name();
+
+	std::ifstream file(file_path, std::ifstream::in);
+	if (file.fail()) {
+		return 0;
+	}
+	std::streampos fileSize;
+	file.seekg(0, std::ios::end);
+	fileSize = file.tellg();
+	file.seekg(0, std::ios::beg);
+
+	contents_.reserve(fileSize);
+	contents_.insert(contents_.begin(),
+					 std::istream_iterator<unsigned char>(file),
+					 {});
+	return 1;
 }
 
-} // end namespace Dbo
-} // end namespace Wt
+git_repository* File_data::repo_init_(std::string repo_path) {
+
+	const char* path = repo_path.c_str();
+	git_repository* repo;
+	if (!git_repository_open_ext(&repo, path, 0, NULL)) {
+		return repo;
+	}
+	
+	git_repository_init(&repo, path, 0);
+	return repo;
+}
+
+bool File_data::repo_add_commit_(git_repository *repo, 
+				const char* repo_path, const char* committer, int update) {
+
+	git_oid commit_oid, tree_oid;
+	git_tree *tree;
+	git_index *index;
+	git_object *parent = NULL;
+	git_reference *ref = NULL;
+	git_signature *signature;
+	git_strarray array;
+	array.strings = (char**) &repo_path;
+	array.count = 1;
+
+	int error;
+
+	error = git_revparse_ext(&parent, &ref, repo, "HEAD");
+	if (error == GIT_ENOTFOUND) {
+		error = 0;
+	} else if (error) {
+		const git_error *err = git_error_last();
+		if (err) std::cout << "Error " << err->klass << " " << err->message << std::endl;
+		return false;
+
+	} else {
+		std::cout << "Head is found. Creating second commit!\n\n" << std::endl;
+	}
+	
+	if (git_repository_index(&index, repo)) {
+		return false;
+	}
+
+	if (!update) {
+
+		if (git_index_add_all(index, NULL, 0, NULL, NULL)){ 
+			return false;
+		}
+	} else {
+		
+		if (git_index_update_all(index, NULL, NULL, NULL)) {
+			return false;
+		}
+	}
+
+	if (git_index_write(index)) {
+		return false;
+	}
+
+	if (git_index_write_tree(&tree_oid, index)) {
+		return false;
+	}
+
+	if (git_tree_lookup(&tree, repo, &tree_oid)) {
+		return false;
+	}
+
+	if (git_signature_now(&signature, committer, committer)) {
+		return false;
+	}
+
+	if (git_commit_create_v(
+							&commit_oid,
+							repo,
+							"HEAD",
+							signature,
+							signature,
+							NULL,
+							"Message",
+							tree,
+							parent ? 1 : 0, parent)) {
+
+		return false;
+	}
+
+	git_index_free(index);
+	git_signature_free(signature);
+	git_tree_free(tree);
+
+	return true;
+}	
+
+std::mutex* File_data::get_mutex_(std::string repo_path) {
+
+	const std::lock_guard<std::mutex> lock(access_global_);
+	
+	if (mtx_map_.count(repo_path)){
+		return mtx_map_[repo_path];
+	}
+	
+	mtx_map_[repo_path] = new std::mutex();
```

Comment:
~~Pretty sure this allocation leaks!~~ Is there a reason why you’re storing `mutex*`s in the map rather than `mutex` values?

Reply:
Actually yeah, std::mutex doesn't have a copy constructor so it won't let me hold a map.